src/java.base/share/classes/java/lang/reflect/code/interpreter/Interpreter.java

```diff
@@ -559,6 +564,25 @@ static Object exec(MethodHandles.Lookup l, OpContext oc, Op o) {
                     .map(oc::getValue)
                     .map(String::valueOf)
                     .collect(Collectors.joining());
+        } else if (o instanceof CoreOp.MonitorOp.MonitorEnterOp) {
```

Comment:
Crazy suggestion: given the recent push of making more of the monitor logic be expressed in Java (as part of Loom), I wonder if it would make sense to imagine a world where a `synchronized` is just lowered to plain Java code - no special core ops? The really fundamental operation that cannot be expressed here is the "gimme a lock for this object". But if we had that (and that is a single core op, or maybe just expressed as a special "field access") wouldn't we be able to achieve the same result?

e.g.

```
lock = lockFor(synchronizedValue)
lock.lock();
try {
    // stats
} finally {
    lock.unlock();
}
```

Reply:
I would very much like to do that, alas we cannot today. `s.m.Unsafe` used to have methods to lock, try-lock, and unlock object monitors. I removed them :-) The JVM permits but does not enforce (HotSpot does enforce) constraints on structured locking (JVM spec section [2.11.10](https://docs.oracle.com/javase/specs/jvms/se22/html/jvms-2.html#jvms-2.11.10))